Lesson_Materials/Lecture_20_More_SYCL_Features/index.html

```diff
@@ -0,0 +1,355 @@
+				if(window.addEventListener) {
+					window.addEventListener('load', () => {
+						let slides = document.getElementsByClassName("slide-background");
+
+						if (slides.length === 0) {
+							slides = document.getElementsByClassName("pdf-page")
+						}
+
+						// Insert global images on each slide
+						for(let i = 0, max = slides.length; i < max; i++) {
+							let cln = document.getElementById("global-images").cloneNode(true);
+							cln.removeAttribute("id");
+							slides[i].appendChild(cln);
+						}
+
+						// Remove top level global images
+						let elem = document.getElementById("global-images");
+						elem.parentElement.removeChild(elem);
+					}, false);
+				}
+			})();
+		</script>
+		<style>
+		/* Custom styling for the highlighted code */
+		.highlight {
+			color: red;
+			font-weight: bold;
+		}
+		</style>
+	</head>
+	<body>
+		<div class="reveal">
+			<div class="slides">
+				<div id="global-images" class="global-images">
+					<img src="../common-revealjs/images/sycl_academy.png" />
+					<img src="../common-revealjs/images/sycl_logo.png" />
+					<img src="../common-revealjs/images/trademarks.png" />
+					<img src="../common-revealjs/images/codeplay.png" />
+				</div>
+				<!--Slide 1-->
+				<section class="hbox">
+					<div class="hbox" data-markdown>
+## More SYCL Features
+					</div>
+				</section>
+				<!--Slide 2-->
+				<section class="hbox" data-markdown>
+## Learning Objectives
+* Learn about atomic operations and how to use them in SYCL kernels
+* Learn about SYCL group algorithms
+* Learn about SYCL reductions
+				</section>
+				<!--Slide 3-->
+				<section>
+					<div class="hbox" data-markdown>
+#### Race Condition
+					</div>
+					<div class="container">
+					<div class="col" data-markdown>
+* In a multithreaded environment, multiple work items writing indiscriminately
+  to the same area of memory causes a race condition.
+					</div>
+					<div class="col" data-markdown>
+```
+q.parallel_for([=](sycl::item&lt;1&gt; it) {
+  // Race condition! Multiple threads
+  // writing to same area of memory
+  a[0] = it.get_global_linear_id();
+});
+```
+					</div>
+				</div>
+				</section>
+				<!--Slide 4-->
+				<section>
+					<div class="hbox" data-markdown>
+#### Race Condition
+					</div>
+					<div class="container">
+						<div class="col" data-markdown>
+![SYCL](./RaceCondition0.png "SYCL")
+						</div>
+						<div class="col" data-markdown>
+```
+q.parallel_for([=](sycl::item&lt;1&gt; it) {
+  // Race condition! Multiple threads
+  // writing to same area of memory
+  a[0] = it.get_global_linear_id();
+});
+```
+* When multiple work items concurrently write different values to the same area
+  of memory, there is no way of knowing which value will be held in memory
+  once all work items have finished writing.
+* This is called a race condition and can be a source of non determinism in
+  code execution.
+* Atomic operations are needed in order to deterministically combine values
```

Comment:
You could maybe add that atomics force a particular ordering of instructions across work-items.

Reply:
Change made